src/kernel/transport/rdma/xio_rdma_datapath.c

```diff
@@ -3070,23 +3078,32 @@ static int xio_rdma_prep_req_in_data(struct xio_rdma_transport *rdma_hndl,
 	}
 	data_len = tbl_length(sgtbl_ops, sgtbl);
 	hdr_len  = vmsg->header.iov_len;
-	if (hdr_len && hdr_len >= rdma_hndl->peer_max_header) {
-		ERROR_LOG("hdr_len=%zd is bigger than peer_max_reader=%d\n",
+	if (hdr_len && hdr_len > rdma_hndl->peer_max_header) {
+		ERROR_LOG("hdr_len=%zd is bigger than peer_max_header=%d\n",
 				hdr_len, rdma_hndl->peer_max_header);
 		return -1;
+	} else if (!hdr_len) {
+		hdr_len = rdma_hndl->peer_max_header;
```

Comment:
the requester can hints the transport about expected header that the peer about to send. this this will allow to utilize the header buffer efficiently. if the user did not provide size, the low level must save the maximum header that peer declared (exchanged upon connection establishment)